lib/config_cat/rollout.ex

```diff
@@ -98,165 +237,338 @@ defmodule ConfigCat.Rollout do
-    hash_candidate = "#{key}#{user_key}"
+    comparison_context = %ComparisonContext{
+      condition: condition,
+      context_salt: context_salt,
+      key: key,
+      salt: salt
+    }
 
-    {hash_value, _} =
-      :sha
-      |> :crypto.hash(hash_candidate)
-      |> Base.encode16()
-      |> String.slice(0, 7)
-      |> Integer.parse(16)
+    comparator = UserCondition.comparator(condition)
+    comparison_value = UserCondition.comparison_value(condition)
 
-    rem(hash_value, 100)
+    case UserComparator.compare(comparator, user_value, comparison_value, comparison_context) do
+      {:ok, result} ->
+        {:ok, result}
+
+      {:error, :invalid_datetime} ->
+        message = "'#{user_value}' is not a valid Unix timestamp (number of seconds elapsed since Unix epoch)"
+        handle_invalid_user_attribute(condition, message, context)
+
+      {:error, :invalid_float} ->
+        message = "'#{user_value}' is not a valid decimal number"
+        handle_invalid_user_attribute(condition, message, context)
+
+      {:error, :invalid_string_list} ->
+        message = "'#{user_value}' is not a valid string array"
+        handle_invalid_user_attribute(condition, message, context)
+
+      {:error, :invalid_version} ->
+        trimmed = user_value |> to_string() |> String.trim()
+        message = "'#{trimmed}' is not a valid semantic version"
+        handle_invalid_user_attribute(condition, message, context)
+    end
+  end
+
+  defp evaluate_segment_condition(condition, %Context{user: nil} = context) do
+    EvaluationWarnings.warn_missing_user(context.warnings, context.key)
+    EvaluationLogger.log_skipping_segment_condition_missing_user(context.logger, condition)
+    {:error, "cannot evaluate, User Object is missing"}
   end
 
-  defp base_value(setting_descriptor, default_value, logs) do
-    result = Map.get(setting_descriptor, Constants.value(), default_value)
-    log(logs, "Returning #{result}")
+  defp evaluate_segment_condition(condition, %Context{} = context) do
+    %Context{logger: logger} = context
+
+    case SegmentCondition.fetch_segment(condition) do
+      {:error, :not_found} ->
+        raise EvaluationError, "Segment reference is invalid."
+
+      {:ok, segment} ->
+        comparator = SegmentCondition.segment_comparator(condition)
+        name = Segment.name(segment)
+
+        EvaluationLogger.log_evaluating_segment_condition_start(logger, condition, name)
+
+        segment
+        |> Segment.conditions()
+        |> Enum.with_index()
+        |> Enum.reduce_while({:ok, true}, fn {condition, index}, acc ->
+          EvaluationLogger.log_evaluating_condition_start(logger, index)
 
-    result
+          result = evaluate_user_condition(condition, name, context)
+          # Faking multiple conditions; may want to use actual condition count
+          # eventually. Keeping it this way to match Python SDK for now.
+          EvaluationLogger.log_evaluating_user_condition_result(logger, result, 2)
```

Comment:
The code referenced above is about how we log the result of evaluating a user condition inside of a segment condition, not about how the segment condition itself is logged. The specific log output is either `=> true` or `=> false, skipping the remaining AND conditions`.

In the Python code that I used as a reference, we have the following cases:

1. When evaluating a top-level user condition, we only print that output if there are >= 1 condition; otherwise nothing is printed.
2. When evaluating a user condition within a segment condition, we print that message regardless of how many conditions there are.

By passing a `2` here, I can reuse the code for printing the message and force it to always print the message even if there is only one condition. That feels a bit ugly, which is why I added the comment.

If that logic is incorrect, let me know how I should fix it.

Reply:
> The code referenced above is about how we log the result of evaluating a user condition inside of a segment condition, not about how the segment condition itself is logged

Sorry, this was an oversight on my part.

> When user object is missing and the targeting rule contains no more conditions besides the segment condition (that is, condition count == 1), we put the THEN part in the same line to make the log more concise (and to be consistent with the way user conditions are logged). Otherwise, we put THEN in a separate line.

This is checked by https://github.com/configcat/elixir-sdk/blob/cc9361b27067fcba92c3b3f4fde5a28267684a83/test/support/fixtures/evaluation/segment/segment_no_user.txt

> When user object is missing, we omit the details of the segment evaluation.

Unfortunately, this case has not been tested so far. But now I specified a test case for this too: https://github.com/configcat/common-js/pull/102/commits/637c5ed8ad9539d074bb656513b4df1a36cf9051

Could you add this to the Elixir SDK's test suite? (And, @kp-cat, also to the Python SDK's?)

Reply:
Done in 00b78c3. The new test is passing, but some others are not yet.